gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java

```diff
@@ -196,31 +121,23 @@ public void close() throws IOException {
       writableByteChannel.close();
     } catch (Exception e) {
       GoogleCloudStorageEventBus.postOnException();
-      throw new IOException(String.format("Upload failed for '%s'", resourceId), e);
+      throw new IOException(
+          String.format("Upload failed for '%s'. reason=%s", resourceId, e.getMessage()), e);
     } finally {
       writableByteChannel = null;
     }
   }
 
-  @Override
-  public void handleResponse(Boolean response) {
-    this.uploadSucceeded = response;
-  }
-
-  @Override
-  protected String getResourceString() {
-    return resourceId.toString();
-  }
-
-  public boolean isUploadSuccessful() {
-    return uploadSucceeded;
-  }
-
   private int writeInternal(ByteBuffer byteBuffer) throws IOException {
     int bytesWritten = writableByteChannel.write(byteBuffer);
     logger.atFinest().log(
         "%d bytes were written out of provided buffer of capacity %d",
         bytesWritten, byteBuffer.limit());
     return bytesWritten;
   }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
```

Comment:
call to storage `writer` is blocking, any other subsequent and may effect the runtime. Have we taken this into account?

Reply:
Even the pipe based implementation will block after the buffer fills up. Since the caller is expected to be much faster than network, most of the time the buffering is not likely to help. In fact based on our benchmarks, the non-pipe based implementation is faster and consumed less memory.